### PR TITLE
Add Hexadecimal parsing for strtof

### PR DESCRIPTION
**Pull request by flyingfish89**
## Summary
solve:
  almost the Hexadecimal string string->float
such as:
  code:float num;
       const char *s= "0x123p32lala";
       char *p;
       num=strtof(s,&p);
       printf("num is %f\n",num);
       printf("str is %s\n",p);
  output:num is 1249835483136.000000
         str is lala

but if the input number is much big;
like:
  code:const char *s2= "0x999999p100";
       num=strtof(s2,&p);
       printf("num is %f\n",num);
       printf("str is %s\n",p);
  corrent : num is 12760587998944832242938906880669384704.000000
  real:     num is 12760587998944800000000000000000000000.000000
it didn't have enough precision
I changed the types of the relevant data from float to double.But it's output changed nothing.
## Impact
add some code
## Testing
 int countnum = 1;
    float num;
    const char *s= "0x123p32";
    char *p; 
    num=strtof(s,&p);
    printf("num is %f\n",num);
    printf("str is %s\n",p);
    printf("count is %d\n",countnum++);
    printf("\n");

    const char *s1= "0x123p32jhdgyeg";
    num=strtof(s1,&p);
    printf("num is %f\n",num);
    printf("str is %s\n",p);
    printf("count is %d\n",countnum++);
    printf("\n");

    const char *s2= "0x999999p100";
    num=strtof(s2,&p);
    printf("num is %f\n",num);
    printf("str is %s\n",p);
    printf("count is %d\n",countnum++);
    printf("\n");

    const char *s3= "0x2.123ppsy32";
    num=strtof(s3,&p);
    printf("num is %f\n",num);
    printf("str is %s\n",p);
    printf("count is %d\n",countnum++);
    printf("\n");

    const char *s4= "0x642.56p90sliu";
    num=strtof(s4,&p);
    printf("num is %f\n",num);
    printf("str is %s\n",p);
    printf("count is %d\n",countnum++);

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@flyingfish89 please fix the warning reported here:
https://github.com/apache/nuttx/actions/runs/3566043017/jobs/5991981398
you can verify it locally:
./tools/checkpath.sh -g HEAD~...HEAD

**Comment by xiaoxiang781216** [comment]
need fix strtod too

File: libs/libc/stdlib/lib_strtof.c [hunk 1]
```diff
@@ -82,6 +83,34 @@ static inline int is_real(float x)
   return (x < infinite) && (x >= -infinite);
 }
 
+int chtof(char c, float base, FAR float *number)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:86]: static bool

File: libs/libc/stdlib/lib_strtof.c [hunk 2]
```diff
@@ -82,6 +83,34 @@ static inline int is_real(float x)
...
+{
+  /* This function is to determine if c is keyword
+   * Then set number + base
+   */
+
+  if (isdigit(c) || (c >= 'A' && c <= 'F')
+                 || (c >= 'a' && c <= 'f'))
+    {
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:94]: remove

File: libs/libc/stdlib/lib_strtof.c [hunk 3]
```diff
@@ -82,6 +83,34 @@ static inline int is_real(float x)
...
+      if (isdigit(c))
+        {
+          *number = *number * base + (double)(c - '0');
+        }
+      else if (c >= 'A' && c <= 'F')
+        {
+          *number = *number * base + (double)(c - '7');
+        }
+      else
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:103]: else fi (c >= 'a' && c <= 'f')

File: libs/libc/stdlib/lib_strtof.c [hunk 4]
```diff
@@ -82,6 +83,34 @@ static inline int is_real(float x)
...
+  /* This function is to determine if c is keyword
+   * Then set number + base
+   */
+
+  if (isdigit(c) || (c >= 'A' && c <= 'F')
+                 || (c >= 'a' && c <= 'f'))
+    {
+      if (isdigit(c))
+        {
+          *number = *number * base + (double)(c - '0');
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:98]: remove all (double) cast

File: libs/libc/stdlib/lib_strtof.c [hunk 5]
```diff
@@ -142,18 +172,37 @@ float strtof(FAR const char *str, FAR char **endptr)
...
+    {
+      while (chtof(*p, p10, &number))
+        {
+          p++;
+          num_digits++;
+        }
+    }
+  else
+    {
+      while (isdigit(*p))
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:201]: let's use the same code as 192 and remove flag_sixt

File: libs/libc/stdlib/lib_strtof.c [hunk 6]
```diff
@@ -82,6 +83,34 @@ static inline int is_real(float x)
   return (x < infinite) && (x >= -infinite);
 }
 
+int chtof(char c, float base, FAR float *number)
+{
+  /* This function is to determine if c is keyword
+   * Then set number + base
+   */
+
+  if (isdigit(c) || (c >= 'A' && c <= 'F')
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:93]:
```
float tmp = base;

if (isdigit(c))
  {
    tmp = c - '0';
  }
else if (islower(c))
  {
    tmp = c - 'a'
  }
else if (isupper(c))
  {
    tmp = c - 'A'
  }

if (c >= base)
  {
    return false;
  }

*number = *number * base + tmp;
return true;
```

File: libs/libc/stdlib/lib_strtof.c [hunk 7]
```diff
@@ -162,13 +211,29 @@ float strtof(FAR const char *str, FAR char **endptr)
...
+          while (chtof(*p, p10, &number))
+            {
+              p++;
+              num_digits++;
+              num_decimals++;
+            }
+      }
+    else
+      {
+        while (isdigit(*p))
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c]: merge else into if

File: libs/libc/stdlib/lib_strtof.c [hunk 8]
```diff
@@ -228,27 +294,53 @@ float strtof(FAR const char *str, FAR char **endptr)
...
     }
 
   if (exponent < __FLT_MIN_EXP__ ||
       exponent > __FLT_MAX_EXP__)
     {
       set_errno(ERANGE);
-      number = infinite;
+      if (exponent < __FLT_MIN_EXP__)
+        {
+          number = divzero;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:327]: why need

File: libs/libc/stdlib/lib_strtof.c [hunk 9]
```diff
@@ -228,27 +294,53 @@ float strtof(FAR const char *str, FAR char **endptr)
...
+
       goto errout;
     }
 
   /* Scale the result */
 
-  p10 = 10.0F;
+  if (flag_sixt == 1)
+    {
+      p10 = 2.0f;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:341]: why divide by 2 not 16

File: libs/libc/stdlib/lib_strtod.c [hunk 10]
```diff
@@ -80,6 +80,36 @@ static inline int is_real(double x)
   return (x < infinite) && (x >= -infinite);
 }
 
+static bool chtod(char c, double base, FAR double *number)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:83]: change line 72 return type to bool

File: libs/libc/stdlib/lib_strtod.c [hunk 11]
```diff
@@ -140,16 +170,22 @@ double strtod(FAR const char *str, FAR char **endptr)
...
+  p10          = 10.;
   number       = 0.;
   exponent     = 0;
   num_digits   = 0;
   num_decimals = 0;
 
   /* Process string of digits */
 
-  while (isdigit(*p))
+  if (*p == '0' && tolower(*(p + 1)) == 'x')
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:181]: add comment:
```
/* Process optional 0x prefix */

if (*p == '0' && tolower(*(p + 1)) == 'x')
{
}

/* Process string of digits */
```

File: libs/libc/stdlib/lib_strtod.c [hunk 12]
```diff
@@ -188,8 +223,16 @@ double strtod(FAR const char *str, FAR char **endptr)
 
   /* Process an exponent string */
 
-  if (*p == 'e' || *p == 'E')
+  if (*p == 'e' || *p == 'E' || *p == 'p' || *p == 'P')
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:226]: does 0x must be paired with 'p'/'P' not 'e'/'E'

File: libs/libc/stdlib/lib_strtof.c [hunk 13]
```diff
@@ -142,16 +174,24 @@ float strtof(FAR const char *str, FAR char **endptr)
...
   number       = 0.0F;
   exponent     = 0;
   num_digits   = 0;
   num_decimals = 0;
 
+  /* Process optional 0x prefix */
+
+  if (*p == '0' && tolower(*(p + 1)) == 'x')
+    {
+          p += 2;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:187]: remove the extra space

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

File: libs/libc/stdlib/lib_strtod.c [hunk 14]
```diff
@@ -140,16 +172,24 @@ double strtod(FAR const char *str, FAR char **endptr)
...
   exponent     = 0;
   num_digits   = 0;
   num_decimals = 0;
 
+  /* Process optional 0x prefix */
+
+  if (*p == '0' && tolower(*(p + 1)) == 'x')
+    {
+      p += 2;
+      p10  = 16.;
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:186]:
```suggestion
      p10 = 16.;
```

File: libs/libc/stdlib/lib_strtod.c [hunk 15]
```diff
@@ -188,8 +227,17 @@ double strtod(FAR const char *str, FAR char **endptr)
...
-  if (*p == 'e' || *p == 'E')
+  if ((p10 == 10. && (*p == 'e' || *p == 'E'))
+     || (p10 == 16. && (*p == 'p' || *p == 'P')))
     {
+      /* if the Hexadecimal system */
+
+      if (p10 == 16.)
+        {
+          exponent *= 4;
+          p10 = 2.0;
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:238]:
```suggestion
          p10 = 2.;
```
Just to follow the same style

File: libs/libc/stdlib/lib_strtof.c [hunk 16]
```diff
@@ -142,16 +174,24 @@ float strtof(FAR const char *str, FAR char **endptr)
       break;
     }
 
+  p10          = 10.0f;
   number       = 0.0F;
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtof.c:178]:
```suggestion
  number       = 0.0f;
```

**Pull request merged**